src/Avalonia.Controls/SelectedItems.cs

```diff
@@ -0,0 +1,49 @@
+﻿// This source file is adapted from the WinUI project.
+// (https://github.com/microsoft/microsoft-ui-xaml)
+//
+// Licensed to The Avalonia Project under MIT License, courtesy of The .NET Foundation.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+#nullable enable
+
+namespace Avalonia.Controls
+{
+    public interface ISelectedItemInfo
+    {
+        public IndexPath Path { get; }
+    }
+
+    internal class SelectedItems<TValue, Tinfo> : IReadOnlyList<TValue>
+        where Tinfo : ISelectedItemInfo
+    {
+        private readonly List<Tinfo> _infos;
+        private readonly Func<List<Tinfo>, int, TValue> _getAtImpl;
+
+        public SelectedItems(
+            List<Tinfo> infos,
+            int count,
+            Func<List<Tinfo>, int, TValue> getAtImpl)
+        {
+            _infos = infos;
+            _getAtImpl = getAtImpl;
+            Count = count;
+        }
+
+        public TValue this[int index] => _getAtImpl(_infos, index);
+
+        public int Count { get; }
+
+        public IEnumerator<TValue> GetEnumerator()
+        {
+            for (var i = 0; i < Count; ++i)
+            {
+                yield return this[i];
```

Comment:
I am already getting flashbacks to visual tree operations using `yield` and taking 10x longer than they should :) Can you just implement a simple struct enumerator?

Reply:
I did consider this, but `SelectedItems` is internal and only exposed to the outside world as an `IReadOnlyList<T>`. We could expose `SelectedItems` but I was trying to stick to WinUI's public API. I think probably the best thing to do here is use the indexer if you want to avoid allocations.

Reply:
Again, might make sense to address perf issues in another PR rather than obfuscate this one.